### PR TITLE
Change Tensor data to shared_ptr

### PR DESCRIPTION
**Pull request by zhoonit**
This PR proposes to change Tensor data to shared_ptr.

Part of #217

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

**Comment by kparichay** [comment]
In my understanding, making tensor data to shared_ptr changes the semantics of the copy constructor of the tensor. 
Without shared_ptr, a new copy of the vector data was being generated (using default copy constructor of vector). However, with shared_ptr, that copy is replaced with just a reference count increment (using default copy constructor of shared_ptr). It reduces memory copies :)
Please check that the usages of tensor in other files can work as per the updated copy constructor.

**Comment by zhoonit** [comment]
> In my understanding, making tensor data to shared_ptr changes the semantics of the copy constructor of the tensor.
> Without shared_ptr, a new copy of the vector data was being generated (using default copy constructor of vector). However, with shared_ptr, that copy is replaced with just a reference count increment (using default copy constructor of shared_ptr). It reduces memory copies :)
> Please check that the usages of tensor in other files can work as per the updated copy constructor.

Ur right. I gonna right a test and check. Thanks :)

File: nntrainer/src/tensor.cpp [hunk 1]
```diff
@@ -63,15 +63,15 @@ static auto rng = [] {
 
 Tensor::Tensor(const TensorDim d) {
   dim = d;
-  this->data = std::vector<float>(dim.getDataLen());
+  this->_data = std::make_shared<std::vector<float>>(dim.getDataLen());
```
Comment (dongju-chae) [inline nntrainer/src/tensor.cpp:66]: This is not related to this PR but I think you can declare and use a base constructor of `Tensor` initializing common variables.

Comment (zhoonit) [reply nntrainer/src/tensor.cpp:66]: That is actually a good idea. Would do in another PR. Thanks for the suggestion!

**Comment by zhoonit** [comment]
> In my understanding, making tensor data to shared_ptr changes the semantics of the copy constructor of the tensor.
> Without shared_ptr, a new copy of the vector data was being generated (using default copy constructor of vector). However, with shared_ptr, that copy is replaced with just a reference count increment (using default copy constructor of shared_ptr). It reduces memory copies :)
> Please check that the usages of tensor in other files can work as per the updated copy constructor.

As this PR is not about changing the interface of Tensor. I am to make this PR keep the copying behavior as before(using copy / move constructor) and deal with the interface later after things are settle down in #185.  @kparichay would that be fine with you?

**Comment by kparichay** [comment]
@zhoonit Please once confirm about this PR with @wooksong before moving forward.

**Comment by zhoonit** [comment]
> @zhoonit Please once confirm about this PR with @wooksong before moving forward.

As far as i understood, his concern was: 
1. copying `vector<T>` is expensive compare to `vector <T *>` 
2. `vector<T *>`, `T *` could be a dangling pointer
3. elements in vector <T> should be destroyed correctly. (I checked and destructor of elements are called when vector is destoryed except the case of `2.` )

I think this should be considered when we expand tensor to another type. 

So far, I think his concern is not related to the PR. This PR aims to make `_data` accessible to other `tensor`.

**Comment by zhoonit** [comment]
There were few concerns that this PR is potentially dangerous, I decided to implement this feature(sharing inner data) in other way.

**Pull request closed without merging**